community/server/src/main/java/org/neo4j/server/rest/security/SecurityEnforcer.java

```diff
@@ -0,0 +1,70 @@
+/*
+ * Copyright (c) 2002-2015 "Neo Technology,"
+ * Network Engine for Objects in Lund AB [http://neotechnology.com]
+ *
+ * This file is part of Neo4j.
+ *
+ * Neo4j is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.server.rest.security;
+
+import org.neo4j.function.Predicate;
+import org.neo4j.function.Predicates;
+
+import javax.servlet.http.HttpServletRequest;
+
+/**
+ * Created by akollegger on 8/23/15.
+ */
+public class SecurityEnforcer implements SecurityRule, ForbiddenSecurityRule {
+
+    private final SecurityRule innerRule;
+
+    private final Predicate<HttpServletRequest> forbiddenPredicate;
+
+    public SecurityEnforcer(final SecurityRule innerRule) {
+        this.innerRule = innerRule;
+
+        if (innerRule instanceof ForbiddenSecurityRule) {
```

Comment:
Rather than having a single wrapper implementation that detects what the inner is, can I suggest an approach I've seen use more commonly, which is to have a factory method that makes the choice of implementation to return? If you had `ForbiddenSecurityRule#fromSecurityRule(SecurityRule rule)`, then it could do the type check and return objects that are already a `ForbiddenSecurityRule` directly without wrapping them, and otherwise return a simple wrapper implementation that returns `false` for `isForbidden`.

You might also want to reconsider the name `ForbiddenSecurityRule`?